Utils/ObjectPool.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace DUCK.Pooling
+{
+	public static class ObjectPool
+	{
+		private interface Pool
+		{
+			Type ObjectType { get; }
+
+			void Add(PoolableObject obj);
+		}
+
+		private class Pool<T> : Pool 
+			where T : PoolableObject
+		{
+			private Queue<T> pooledObjects = new Queue<T>();
+
+			public Type ObjectType
+			{
+				get
+				{
+					return typeof(T);
+				}
+			}
+
+			public T Get()
+			{
+				var obj = (pooledObjects.Count > 0)
+					? pooledObjects.Dequeue()
+					: null;
```

Comment:
Might be worth considering letting the pool manage object creation as well since it handles the other end of the lifecycle and would be much more convenient for the end user.

Its not unusual for pools to return a new instance of the object where you have `null` if you think that would be less intuitive then a `GetOrCreate()` method would work.

Reply:
It does - ObjectPool is a private class to this will never be called by the end user. See thw two static Instantiate<T> functions - is that not what you mean? They both perform a 'create or retrieve' step